### PR TITLE
JIT: Suppress emitting same-reg zero extending move

### LINKED ISSUE #21923: Unnecessary mov generated in MemoryMarshal.Cast
This:
```C#
[MethodImpl(MethodImplOptions.NoInlining)]
public static ReadOnlySpan<byte> GetBytes(ReadOnlySpan<int> s) => MemoryMarshal.Cast<int, byte>(s);
```
results in this asm:
```
G_M23838_IG01:
       4883EC28             sub      rsp, 40
       90                   nop

G_M23838_IG02:
       488B02               mov      rax, bword ptr [rdx]
       8B5208               mov      edx, dword ptr [rdx+8]
       8BD2                 mov      edx, edx
       48C1E202             shl      rdx, 2
       4881FAFFFFFF7F       cmp      rdx, 0x7FFFFFFF
       770E                 ja       SHORT G_M23838_IG04
       488901               mov      bword ptr [rcx], rax
       895108               mov      dword ptr [rcx+8], edx
       488BC1               mov      rax, rcx

G_M23838_IG03:
       4883C428             add      rsp, 40
       C3                   ret

G_M23838_IG04:
       E83384C95F           call     CORINFO_HELP_OVERFLOW
       CC                   int3
```
but that `mov edx, edx` is superfluous.

@AndyAyersMS is investigating a fix.

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Add a peephole optimization to suppress emitting zero extending moves
if the previous instruction has already done a suitable zero extension.

Only implemented for x64 currently.

Closes #21923

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@BruceForstall PTAL
cc @dotnet/jit-contrib 

Jit diffs shows:
```
PMI Diffs for System.Private.CoreLib.dll, framework assemblies for x64 default jit
Summary:
(Lower is better)
Total bytes of diff: -3488 (-0.01% of base)
    diff is an improvement.
Top file improvements by size (bytes):
       -1192 : System.Private.CoreLib.dasm (-0.03% of base)
        -348 : System.Memory.dasm (-0.22% of base)
        -228 : System.Net.Http.dasm (-0.04% of base)
        -208 : System.Net.NetworkInformation.dasm (-0.45% of base)
        -168 : System.Runtime.Numerics.dasm (-0.23% of base)
51 total files with size differences (51 improved, 0 regressed), 78 unchanged.
Top method improvements by size (bytes):
         -44 (-0.62% of base) : System.Memory.dasm - ReadOnlySequenceDebugView`1:.ctor(struct):this (5 methods)
         -36 (-0.94% of base) : System.Memory.dasm - SequenceReader`1:GetNextSpan():this (4 methods)
         -32 (-0.85% of base) : System.Memory.dasm - SequenceReader`1:ResetReader():this (4 methods)
         -32 (-0.72% of base) : System.Memory.dasm - SequenceReader`1:IsNextSlow(struct,bool):bool:this (4 methods)
         -31 (-0.90% of base) : System.Memory.dasm - BuffersExtensions:CopyToMultiSegment(byref,struct) (5 methods)
Top method improvements by size (percentage):
          -2 (-20.00% of base) : System.Private.CoreLib.dasm - UInt32:System.IConvertible.ToInt64(ref):long:this
          -2 (-20.00% of base) : System.Private.CoreLib.dasm - UInt32:System.IConvertible.ToUInt64(ref):long:this
          -4 (-18.18% of base) : System.IO.FileSystem.dasm - FILE_TIME:ToTicks():long:this
          -2 (-18.18% of base) : System.Net.HttpListener.dasm - HttpListenerTimeoutManager:get_MinSendBytesPerSecond():long:this
          -2 (-18.18% of base) : System.Net.NetworkInformation.dasm - SystemIcmpV4Statistics:get_MessagesSent():long:this
868 total methods with size differences (868 improved, 0 regressed), 192460 unchanged.
```
sample diff (on example from #21923)
```diff
 G_M29305_IG02:
        mov      rax, bword ptr [rdx]
        mov      edx, dword ptr [rdx+8]
-       mov      edx, edx
        shl      rdx, 2
        cmp      rdx, 0xD1FFAB1E
        ja       SHORT G_M29305_IG04
```

**Comment by benaadams** [comment]
Probably fixes https://github.com/dotnet/coreclr/issues/17963 also?

**Comment by benaadams** [comment]
Although that might need a double look back :-/

**Comment by AndyAyersMS** [comment]
We'd get 2 out of 3, not too bad.
```asm
 movzx  r8d,word ptr [rax]  
 add    rax,2  
 mov    r8d,r8d                 ; pointless mov
```
Looking back further is certainly an option. The first prototype (which was almost certainly a bit too aggressive) was getting about 4x as many cases -- it assumed the upstream producer would properly zero extend.

File: src/jit/codegenxarch.cpp [hunk 1]
```diff
@@ -3112,7 +3112,7 @@ void CodeGen::genCodeForCpBlkRepMovs(GenTreeBlk* cpBlkNode)
     else
 #endif
     {
-#ifdef _TARGET_X64_
+#ifdef _TARGET_AMD64_
```
Comment (BruceForstall) [inline src/jit/codegenxarch.cpp]: This looks unrelated; separate PR?

File: src/jit/codegenxarch.cpp [hunk 2]
```diff
@@ -6426,6 +6428,14 @@ void CodeGen::genIntToIntCast(GenTreeCast* cast)
                 break;
 #ifdef _TARGET_64BIT_
             case GenIntCastDesc::ZERO_EXTEND_INT:
+                // We can skip emitting this zero extending move if the previous instruction zero extended implicitly
+                if ((srcReg == dstReg) && (emit->emitCurIGinsCnt > 0) && compiler->opts.OptimizationEnabled())
```
Comment (BruceForstall) [inline src/jit/codegenxarch.cpp]: I'm not fond of leaking `emitLastIns` out of the emitter. Maybe this should call a function like emitIsLastInsCall(), say `emitIsLastInsZeroExtendingWrite(srcReg)`. Then this code could be:
```
canSkip = (srcReg == dstReg) && compiler->opts.OptimizationEnabled() && emit->emitIsLastInsZeroExtendingWrite(srcReg);
```

**Comment by stephentoub** [comment]
Thanks, Andy.

File: src/jit/emitxarch.cpp [hunk 3]
```diff
@@ -145,6 +145,31 @@ bool emitter::IsDstSrcSrcAVXInstruction(instruction ins)
     return ((CodeGenInterface::instInfo[ins] & INS_Flags_IsDstSrcSrcAVXInstruction) != 0) && IsAVXInstruction(ins);
 }
 
+bool emitter::doesZeroExtendingWrite(instrDesc* id, regNumber reg)
```
Comment (BruceForstall) [inline src/jit/emitxarch.cpp]: Needs a function header comment

File: src/jit/codegenxarch.cpp (hunk 1 above)
Comment (mikedn) [inline src/jit/codegenxarch.cpp]: What's up with this change?

Comment (AndyAyersMS) [reply src/jit/codegenxarch.cpp]: Forgot that was in there.

File: src/jit/codegenxarch.cpp (hunk 2 above)
Comment (AndyAyersMS) [reply src/jit/codegenxarch.cpp]: Sure, let me update.

File: src/jit/emitxarch.cpp [hunk 4]
```diff
@@ -145,6 +145,31 @@ bool emitter::IsDstSrcSrcAVXInstruction(instruction ins)
...
+        case IF_RRW_CNS:
+        case IF_RRW_SHF:
+        case IF_RWR_RRD:
+        case IF_RRW_RRD:
+        case IF_RWR_MRD:
+        case IF_RWR_SRD:
+        case IF_RWR_ARD:
+
+            // Can't rely on a "small" movsx as we will over-extend to 8 bytes
+            return (id->idIns() != INS_movsx) && (id->idReg1() == reg) && (id->idOpSize() != EA_8BYTE);
```
Comment (mikedn) [inline src/jit/emitxarch.cpp]: The size check doesn't seem right - x64 only zero extends 32 bit register writes, `mov al, 42` does not zero extend. So the check should be ` == EA_4BYTE`, though that will probably require special casing for `movzx` where size has a different meaning.

File: src/jit/codegenxarch.cpp (hunk 1 above)
Comment (AndyAyersMS) [reply src/jit/codegenxarch.cpp]: Sure.

File: src/jit/codegenxarch.cpp (hunk 2 above)
Comment (mikedn) [reply src/jit/codegenxarch.cpp]: `emitAreUpper32bitZero` might be better, it doesn't matter how they ended up being zero, just that they are.

File: src/jit/emitxarch.cpp (hunk 4 above)
Comment (AndyAyersMS) [reply src/jit/emitxarch.cpp]: Good point.

For `movzx` you mean allow both 4 and 8 byte sizes...?

File: src/jit/codegenxarch.cpp (hunk 2 above)
Comment (AndyAyersMS) [reply src/jit/codegenxarch.cpp]: Agree -- I'll go with that.

File: src/jit/emitxarch.cpp (hunk 4 above)
Comment (mikedn) [reply src/jit/emitxarch.cpp]: For `movzx` and `movsx` the size indicates the source operand size - 1 or 2 bytes - rather than the destination operand size. So `movzx` always zeroes out the upper 32 bit, no matter the size.

It's a bit unfortunate. I'm considering changing movzx/movsx to movzxb/movzxw/movsxb/movsxw in the future, that might simplify things and perhaps allow more code to be shared with ARM.

**Comment by AndyAyersMS** [comment]
Updated. Generates same diffs as the first version.

File: src/jit/codegenxarch.cpp (hunk 1 above)
Comment (AndyAyersMS) [reply src/jit/codegenxarch.cpp]: Removed it from this PR.

**Comment by AndyAyersMS** [comment]
OSX jenkins hangup, retrying

@dotnet-bot retest OSX10.12 x64 Checked Innerloop Build and Test
